core/server/worker/src/main/java/alluxio/worker/grpc/AbstractWriteHandler.java

```diff
@@ -79,66 +93,107 @@
    * @param writeRequest the request from the client
    */
   public void write(WriteRequest writeRequest) {
-    try {
-      if (mContext == null) {
-        LOG.debug("Received write request {}.", writeRequest);
-        mContext = createRequestContext(writeRequest);
-      } else {
-        Preconditions.checkState(!mContext.isDoneUnsafe(),
-            "invalid request after write request is completed.");
-      }
-      validateWriteRequest(writeRequest);
-      if (writeRequest.hasCommand()) {
-        WriteRequestCommand command = writeRequest.getCommand();
-        if (command.getFlush()) {
-          flush();
+    if (!tryAcquireSemaphore()) {
+      return;
+    }
+    mSerializingExecutor.execute(() -> {
```

Comment:
What will happen if client tries zero copy but the server does not support it?

Reply:
The zero copy code path serialize and deserialize messages in the same format as the protobuf marshaller, therefore the client and the server can switch between zero copy or protobuf marshaller without breaking compatibility with the other side.